firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteIndexManager.java

```diff
@@ -87,6 +108,259 @@ public void addFieldIndex(FieldIndex index) {
+          "Expected upper and lower bound size to match");
+
+      // TODO(indexing): To avoid reading the same documents multiple times, we should ideally only
+      // send one query that combines all clauses.
+      for (int i = 0; i < lowerBoundValues.size(); ++i) {
+        db.query(
+                String.format(
+                    "SELECT document_name from index_entries WHERE index_id = ? AND index_value %s ? AND index_value %s ?",
+                    lowerBound.isBefore() ? ">=" : ">", upperBound.isBefore() ? "<=" : "<"))
+            .binding(fieldIndex.getIndexId(), lowerBoundValues.get(i), upperBoundValues.get(i))
+            .forEach(
+                row -> result.add(DocumentKey.fromPath(ResourcePath.fromString(row.getString(0)))));
+      }
+    } else {
+      List<byte[]> lowerBoundValues =
+          encodeTargetValues(fieldIndex, target, lowerBound.getPosition());
+      for (byte[] lowerBoundValue : lowerBoundValues) {
+        db.query(
+                String.format(
+                    "SELECT document_name from index_entries WHERE index_id = ? AND index_value %s  ?",
+                    lowerBound.isBefore() ? ">=" : ">"))
+            .binding(fieldIndex.getIndexId(), lowerBoundValue)
+            .forEach(
+                row -> result.add(DocumentKey.fromPath(ResourcePath.fromString(row.getString(0)))));
+      }
+    }
+
+    Logger.debug(TAG, "Index scan returned %s documents", result.size());
+    return result;
+  }
+
+  /**
+   * Returns an index that can be used to serve the provided target. Returns {@code null} if no
+   * index is configured.
+   */
+  private @Nullable FieldIndex getMatchingIndex(Target target) {
+    TargetIndexMatcher targetIndexMatcher = new TargetIndexMatcher(target);
+    String collectionGroup =
+        target.getCollectionGroup() != null
+            ? target.getCollectionGroup()
+            : target.getPath().getLastSegment();
+
+    List<FieldIndex> activeIndices = new ArrayList<>();
+
+    db.query(
+            "SELECT index_id, index_proto FROM index_configuration WHERE collection_group = ? AND active = 1")
+        .binding(collectionGroup)
+        .forEach(
+            row -> {
+              try {
+                FieldIndex fieldIndex =
+                    serializer.decodeFieldIndex(
+                        collectionGroup, row.getInt(0), Index.parseFrom(row.getBlob(1)));
+                boolean matches = targetIndexMatcher.servedByIndex(fieldIndex);
+                if (matches) {
+                  activeIndices.add(fieldIndex);
+                }
+              } catch (InvalidProtocolBufferException e) {
+                throw fail("Failed to decode index: " + e);
+              }
+            });
+    ;
+
+    if (activeIndices.isEmpty()) {
+      return null;
+    }
+
+    // Return the index with the most number of segments
+    Collections.sort(activeIndices, (l, r) -> Integer.compare(r.segmentCount(), l.segmentCount()));
```

Comment:
Can we use `Collections.max` instead?

Reply:
Probably not since it is likely not available at our API level.

Checking...

Wow, I stand corrected. Java 7 for the win!